bash-preexec.sh

```diff
@@ -244,32 +258,44 @@ __bp_preexec_invoke_exec() {
         return
     fi
 
-    # Invoke every function defined in our function array.
+    local preexec_ret_value
+    bash_preexec_invoke_preexec_functions "${__bp_last_ret_value:-}" "$__bp_last_argument_prev_command" "$this_command"
+
+    # Restore the last argument of the last executed command, and set the return
+    # value of the DEBUG trap to be the return code of the last preexec function
+    # to return an error.
+    # If `extdebug` is enabled a non-zero return value from any preexec function
+    # will cause the user's command not to execute.
+    # Run `shopt -s extdebug` to enable
+    __bp_set_ret_value "$preexec_ret_value" "$__bp_last_argument_prev_command"
+}
+
+# This function invokes every function defined in our function array
+# "preexec_function".  This function receives the arguments $1 and $2 for $?
+# and $_, respectively, that will be set for the preexec functions.  The third
+# argument $3 specifies the user command that is going to be executed
+# (corresponding to BASH_COMMAND in the DEBUG trap).  This function assigns the
+# last non-zero exit status from the preexec functions to the variable
+# `preexec_ret_value`.  If there is no error, preexec_ret_value is set to `0`.
+bash_preexec_invoke_preexec_functions() {
+    local __lastexit=$1 __lastarg=$2 __this_command=$3
```

Comment:
OK, I will follow the existing local variables used.

I have prefixed these new variables with underscores in case the hook functions want to modify the global variables of the conflicting names, but if we would care about it, the same would also apply to the existing variable names. I decided to just follow the existing ones.

Reply:
As I have commented in https://github.com/rcaloras/bash-preexec/pull/129#discussion_r859137813, these local variables can affect the context of callers (i.e., the hook functions stored in the array `preexec_functions` by the user).

Reply:
dda9193